docs: add remote module to docs/tutorial/security.md (#17480)
electron/electron: docs/tutorial/security.md

your code. As such, it is your responsibility to follow a few important best
practices:

> 

Author: even though I don't like trailing whitespace, we should not include this change

Reviewer: Restored.

shell.openExternal('https://example.com/index.html')
```

## 15) Disable the `remote` module

> The `remote` module provides a simple way to do inter-process communication (IPC) between the renderer process (web page) and the main process. Using it, a renderer can invoke methods of a main process object without explicitly sending inter-process messages. If your desktop application does not run untrusted content, this is a very elegant way to have your renderer processes access and work with modules that are only available to the main process, such as GUI-related modules (dialog, menu etc.).

Author: provides a simple way for the renderer processes to access APIs normally only available in the main process

Author: I would also omit web page, it can just be the preload script calling those APIs

Reviewer: Copied this from the `remote` documentation 😁 . But I agree with your changes, amended.

> However, if your app can run untrusted content and you sandbox your renderer processes accordingly, the `remote` module makes it easy for malicious code to escape the sandbox and have access to system resources via the higher privileges of the main process. Therefore, it should be disabled in such circumstances.

Author: **even if** you sandbox your renderer processes

Reviewer: Excellent point, fixed.

### Why?

`remote` uses an internal IPC channel to communicate with the main process. "Prototype pollution" attacks can grant malicious code access to the internal IPC channel, which can then be used to escape the sandbox by mimicking `remote` IPC messages and getting access to main process modules running with higher privileges.

Additionally, if you're not careful, preload scripts may leak modules to a sandboxed renderer. Leaking `remote` arms malicious code with a multitude of main process modules with which to perform an attack..

Disabling the `remote` module eliminates these attack vectors. Enabling context isolation also prevents the "prototype pollution" attacks from succeeding.

### How?

Sandboxed renderers

```js
// Bad
const mainWindow = new BrowserWindow({})
```

```js
// Good
const mainWindow = new BrowserWindow({
  webPreferences: {
    enableRemoteModule: false
  }
})
```

```html
<!-- Bad -->
<webview src="page.html"></webview>

<!-- Good -->
<webview enableremotemodule="false" src="page.html"></webview>
```

## 16) Filter the `remote` module

> If you cannot disable the `remote` module, you should filter the `remote` modules that your application does not require. This can be done by blocking certain modules entirely and by replacing others with proxies that expose only the functionality that your app needs.

Author: you should filter the globals, node and electron modules (so-called built-ins) accessible via `remote`

Reviewer: 👍🏻

shell.openExternal('https://example.com/index.html')
```

## 15) Disable the `remote` module

The `remote` module provides a simple way for the renderer processes to

> access APIsnormally only available in the main process. Using it, a

Reviewer 2: ```suggestion
access APIs normally only available in the main process. Using it, a
```

renderer can invoke methods of a main process object without explicitly sending
inter-process messages. If your desktop application does not run untrusted
content, this is a very elegant way to have your renderer processes access and
work with modules that are only available to the main process, such as

> GUI-related modules (dialog, menu etc.).

Reviewer 2: ```suggestion
GUI-related modules (dialogs, menus, etc.).
```

However, if your app can run untrusted content and even if you
[sandbox][sandbox] your renderer processes accordingly, the `remote` module
makes it easy for malicious code to escape the sandbox and have access to
system resources via the higher privileges of the main process. Therefore,
it should be disabled in such circumstances.

### Why?

`remote` uses an internal IPC channel to communicate with the main process.
"Prototype pollution" attacks can grant malicious code access to the internal
IPC channel, which can then be used to escape the sandbox by mimicking `remote`
IPC messages and getting access to main process modules running with higher
privileges.

Additionally, if you're not careful, preload scripts may leak modules to a
sandboxed renderer. Leaking `remote` arms malicious code with a multitude
of main process modules with which to perform an attack.

Disabling the `remote` module eliminates these attack vectors. Enabling
context isolation also prevents the "prototype pollution" attacks from
succeeding.

### How?

> Sandboxed renderers

Author: removed

```js

> // Bad

Reviewer 2: The bad/good examples make sense here & in the rest of the docs, but I do suspect a lot of people will ignore the text and skip straight down to this.

For that reason, I might modify the comment a little:

```suggestion
// Bad if the renderer can run untrusted content
```

const mainWindow = new BrowserWindow({})
```

```js
// Good
const mainWindow = new BrowserWindow({
  webPreferences: {
    enableRemoteModule: false
  }
})
```

```html

> <!-- Bad -->

Reviewer 2: ```suggestion
<!-- Bad if the renderer can run untrusted content  -->
```

<webview src="page.html"></webview>

<!-- Good -->
<webview enableremotemodule="false" src="page.html"></webview>
```

## 16) Filter the `remote` module

If you cannot disable the `remote` module, you should filter the globals,

> node and electron modules (so-called built-ins) accessible via `remote`

Reviewer 2: ```suggestion
Node, and Electron modules (so-called built-ins) accessible via `remote`
```

malicious code running in a compromised renderer process. By limiting
the set of accessible modules to the minimum that your app needs and
filtering out the others, you reduce the toolset that malicious code
can use to attack the system. Note that you'll still be vulnerable to

> "prototype pollution" attacks on the remote module internal mechanism,

Reviewer 3: Hm, to me this implies that there are known prototype pollution bugs in the remote module mechanism, which isn't the case. Perhaps:

> Note that the safest option is to [fully disable the remote module](#15-disable-the-remote-module). If you choose to filter access rather than completely disable the module, you must be very careful to ensure that no escalation of privilege is possible through the modules you allow past the filter.

Reviewer: Agreed, using now your suggestion.
